### PR TITLE
Select root folder for timeline

### LINKED ISSUE #141: Limit Photos to a single folder + subtree
I think it would be really nice to create an option where the user can select a single folder (like `/Photos`) and the app only scans this folder and its subfolders.

I  have tons and tons of images all over my nextcloud, but only in `/Photos` are my, well, photos. And it's kind of a tedious task to create always a `.nomedia` file in all other directories.

### PR DESCRIPTION
**Pull request by Mikescops**
Closes #141

### CODE REVIEW

**Comment by Mikescops** [comment]
@skjnldsv I don't know why but .jpg (which are JPEG files) are not taken in account, for some reasons, and as of now i can't find why..

**Comment by Mikescops** [comment]
Seems like the dav:or is not working and only the first mime type is taken in account.

**Comment by strugee** [comment]
This only provides for one root folder, right? That would be a problem for me and I assume other folks too since there's a couple places photos go if you use the out-of-the-box defaults - brand new accounts have a "Photos" folder created for them, so that's a natural place to put them, but the Android app auto uploads photos to the "InstantUpload" folder by default.

**Comment by ntimo** [comment]
But when there is no photos directory select manually the app should function as it does with the default settings, I would assume. So this would also some your issue @strugee. Am I correct @Mikescops? :)

**Comment by strugee** [comment]
@ntimo right, that would be a workaround. But then I have the original issue described in #141 again.

**Comment by Mikescops** [comment]
By default no difference with the current version, but you can select one folder from where images on your timeline will be fetched from.

File: src/services/PhotoSearch.js [hunk 1]
```diff
@@ -68,6 +70,15 @@ export default async function(onlyFavorites = false, options = {}) {
...
 		: ''
 
+	const eqOwner = options.rootFolder === ''
+		? `<d:eq>
+				<d:prop>
+					<oc:owner-id/>
+				</d:prop>
+				<d:literal>${getCurrentUser().uid}</d:literal>
+			</d:eq>`
+		: ''
```
Comment (skjnldsv) [inline src/services/PhotoSearch.js:80]: This is needed to not have the shares as results.
Why did you need to put a condition on it?
It should not cause any issues like it was before :)

File: lib/Controller/PageController.php [hunk 2]
```diff
@@ -86,7 +86,10 @@ public function index(): TemplateResponse {
 		$this->initialStateService->provideInitialState($this->appName, 'image-mimes', Application::IMAGE_MIMES);
 		$this->initialStateService->provideInitialState($this->appName, 'video-mimes', Application::VIDEO_MIMES);
 		$this->initialStateService->provideInitialState($this->appName, 'maps', $this->appManager->isEnabledForUser('maps') === true);
+
+		// User Settings
 		$this->initialStateService->provideInitialState($this->appName, 'croppedLayout', $this->config->getUserValue($user->getUid(), Application::APP_ID, 'croppedLayout', 'false'));
+		$this->initialStateService->provideInitialState($this->appName, 'timelineRootFolder', $this->config->getUserValue($user->getUid(), Application::APP_ID, 'timelineRootFolder', '/'));
```
Comment (skjnldsv) [inline lib/Controller/PageController.php:92]: We should probably limit the allowed keys here: https://github.com/nextcloud/photos/blob/441f97020bee10aeb9e0b27cd2ac174ed02dc3b1/lib/Controller/ApiController.php#L66

As a safety mesure :thinking:

File: src/services/PhotoSearch.js (hunk 1 above)
Comment (Mikescops) [reply src/services/PhotoSearch.js:80]: The owner filter is only allowed on the root. Dav throws this error if you try on subfolders. What do you think then?

Comment (skjnldsv) [reply src/services/PhotoSearch.js:80]: Then it should be fixed on server.
cc @rullzer and @icewind1991 :)

Comment (skjnldsv) [reply src/services/PhotoSearch.js:80]: https://github.com/nextcloud/server/pull/17941

Comment (skjnldsv) [reply src/services/PhotoSearch.js:80]: https://github.com/nextcloud/server/blob/2a054e6c04e0a40421510eb889cbf59f153c5177/lib/private/Files/Node/Folder.php#L246

File: lib/Controller/PageController.php (hunk 2 above)
Comment (Mikescops) [reply lib/Controller/PageController.php:92]: I'm not sure what you mean

Comment (skjnldsv) [reply lib/Controller/PageController.php:92]: setUserConfig allows anything, and will put it straight into the db.
Meaning we could override existing values. We should check if the $key is either `timelineRootFolder` or `croppedLayout` and throw otherwise as a safety measure

Comment (Mikescops) [reply lib/Controller/PageController.php:92]: ok gotcha

**Comment by strugee** [comment]
> By default no difference with the current version, but you can select one folder from where images on your timeline will be fetched from.

@Mikescops right, to clarify, I'm not suggesting this would break existing workflows - I'm just saying that the feature might not actually be complete enough for many(?) users due to defaults in the server/Android app. Speaking only for myself, for example, I would not be able to use this feature since given the way my Nextcloud is organized, if I turned it on I would end up missing a ton of images that I'd want to see in the Photos app.

@jancborchardt I'd be interested in hearing if you agree or if you think it's not worth the additional interface/code complexity.

**Comment by Mikescops** [comment]
It's outside of the scope of this PR, if you're not happy with it, open a separate issue after it's merged, thanks.

**Comment by saywebsolutions** [comment]
Which release will this go into? I'm looking forward to being able to use photos again, thank you @Mikescops

**Comment by Mikescops** [comment]
NC 21 at least, for now we're stuck with a server issue.

**Comment by StarSmasher44** [comment]
Hi! Sorry to bring this back, but I am hopeful this merge will resolve my issue. Is there any status update on this?
 I saw the 'server problem' pull had been merged, is there anything else that needs to be done or fixed in order to get this in asap?

**Comment by skjnldsv** [comment]
> I saw the 'server problem' pull had been merged, is there anything else that needs to be done or fixed in order to get this in asap?

Hey! The server hasn't received anything regarding this fix yet. Unfortunately this is not yet possible :/

**Comment by StarSmasher44** [comment]
> > I saw the 'server problem' pull had been merged, is there anything else that needs to be done or fixed in order to get this in asap?
> 
> Hey! The server hasn't received anything regarding this fix yet. Unfortunately this is not yet possible :/

Ah well that sucks, I was referring to this one: https://github.com/nextcloud/server/pull/17941 mentioned some posts up.

But would you happen to know what still needs to be done/added/fixed to make this work? I have a limited background in web development languages so I might be able to pull something through, even though I highly doubt it.

**Comment by skjnldsv** [comment]
> But would you happen to know what still needs to be done/added/fixed to make this work? I have a limited background in web development languages so I might be able to pull something through, even though I highly doubt it.

It require a fair amount of rework on the backend dav search implementation :/

**Comment by StarSmasher44** [comment]
> > But would you happen to know what still needs to be done/added/fixed to make this work? I have a limited background in web development languages so I might be able to pull something through, even though I highly doubt it.
> 
> It require a fair amount of rework on the backend dav search implementation :/

Sorry for being overly asking questions or coming off as pushy in any regard, I am merely just trying to figure out a good entry point, I looked through the code twice but I can't find an obvious stopping point beside the mentioned error with single vs multiple mime types being used.

And I am unsure if you even know enough of this PR yourself to respond (It isn't easy going off using other people's work as a starting point for yourself, imo) but I still have no idea what is actually wrong with this code, what needs fixing and what doesn't.

To me it looks mostly fine, given all it really needs to do is enter/search a single folder and mark that as the 'Photos' folder for NextCloud, if I get this correct.

If I have more information I may try to play around with this to see if I can bring it to a working state somehow, I have a bad taste for using OneDrive as my image backup server ever since Samsung forced it upon me, and this still adds in 600 images from my several Unity projects.

**Comment by skjnldsv** [comment]
@StarSmasher44 I suggest you read the original issues like https://github.com/nextcloud/server/pull/17941 https://github.com/nextcloud/photos/issues/141

Again, this requires a fix on server. Currently there is a hard limitation for our webdav SEARCH that forces you to limit the search to a user home for performance issues.
That is why we don't have external storage nor groupfolders. And if you ignore this, you will not be able to properly filter it anymore. This requires a big revamp for al the following points to be implemented:
- Allow to search/filter by tags (favourites)
- Allow to search/filter by mime
- Allow to search/filter by user
- Allow to search/filter within a folder OR from root (currently only root is allowed with filtering)
- Allow to limit results
- Allow to sort results

Currently, the combo `search-from-folder` AND `filter-by-owner` is missing. It's not enabled (hard-coded on server) Once it is here, we'll be able to tell the Photos app to search from within the provided root folder instead of the default root.

**Comment by timsayshey** [comment]
A rewrite sounds like a long term goal. Should we just stick with the original Gallery plugin until then? I think most people's Photo apps are a mix-mash of random images from all over their cloud.

> @StarSmasher44 I suggest you read the original issues like [nextcloud/server#17941](https://github.com/nextcloud/server/pull/17941) #141 entirely (with the folded comments too)
> 
> Again, this requires a fix on server. Currently there is a hard limitation for our webdav SEARCH that forces you to limit the search to a user home for performance issues.
> That is why we don't have external storage nor groupfolders. And if you ignore this, you will not be able to properly filter it anymore. This requires a big revamp for al the following points to be implemented:
> 
> * Allow to search/filter by tags (favourites)
> * Allow to search/filter by mime
> * Allow to search/filter by user
> * Allow to search/filter within a folder OR from root (currently only root is allowed with filtering)
> * Allow to limit results
> * Allow to sort results
> 
> Currently, the combo `search-from-folder` AND `filter-by-owner` is missing. It's not enabled (hard-coded on server) Once it is here, we'll be able to tell the Photos app to search from within the provided root folder instead of the default root.

**Comment by AndyScherzinger** [comment]
@artonge I guess this PR is obsolete with #2319 having been merged?
